flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java

```diff
@@ -0,0 +1,255 @@
+            throw new FlinkRuntimeException(
+                    String.format(
+                            "Could not create the checkpoint directory '%s'",
+                            directoryForChangelog));
+        }
+
+        return directoryForChangelog;
+    }
+
+    private void deleteMaterialization(LongPredicate pruningChecker) {
+        final Set<Long> materializationToRemove = new HashSet<>();
+        synchronized (lock) {
+            Iterator<Entry<Long, Long>> iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                long entryCheckpointId = entry.getKey();
+                if (pruningChecker.test(entryCheckpointId)) {
+                    materializationToRemove.add(entry.getValue());
+                    iterator.remove();
+                }
+            }
+
+            iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                materializationToRemove.remove(entry.getValue());
+            }
+        }
+
+        for (Long materializationId : materializationToRemove) {
+            File materializedDir =
+                    localRecoveryConfig
+                            .getLocalStateDirectoryProvider()
+                            .orElseThrow(
+                                    () ->
+                                            new IllegalStateException(
+                                                    "Local recovery must be enabled."))
+                            .subtaskSpecificCheckpointDirectory(materializationId);
+
+            if (!materializedDir.exists()) {
+                continue;
+            }
+            try {
+                deleteDirectory(materializedDir);
+            } catch (IOException ex) {
+                LOG.warn(
+                        "Exception while deleting local state directory of materialized part {} in subtask ({} - {} - {}).",
+                        materializedDir,
+                        jobID,
+                        jobVertexID,
+                        subtaskIndex,
+                        ex);
+            }
+        }
+    }
+
+    @Override
+    public void confirmCheckpoint(long confirmedCheckpointId) {
+        // Scenarios:
+        //   c1,m1
+        //   confirm c1, do nothing.
+        //   c2,m1
+        //   confirm c2, delete c1, don't delete m1
+        //   c3,m2
+        //   confirm c3, delete c2, delete m1
+        LOG.debug(
+                "Received confirmation for checkpoint {} in subtask ({} - {} - {}). Starting to prune history.",
+                confirmedCheckpointId,
+                jobID,
+                jobVertexID,
+                subtaskIndex);
+        // delete changelog-chk
+        pruneCheckpoints(checkpointId -> checkpointId < confirmedCheckpointId, false);
```

Comment:
Parent method has `true` here (break iteration).
Is it intentional?

Reply:
Fixed.  It was a mistake, but not break correctness. I think `breakOnceCheckerFalse` is used to reduce the number of iterations.